core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,394 @@
+     * method was called.
+     */
+    private final AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer;
+
+    /**
+     * Set to true once {@code .prepare()} is called on the upstreamResponseTransformer
+     */
+    private final AtomicBoolean preparedCalled = new AtomicBoolean(false);
+
+    /**
+     * Set to true once {@code .onResponse()} is called on the upstreamResponseTransformer
+     */
+    private final AtomicBoolean onResponseCalled = new AtomicBoolean(false);
+
+    /**
+     * Set to true once {@code .onStream()} is called on the upstreamResponseTransformer
+     */
+    private final AtomicBoolean onStreamCalled = new AtomicBoolean(false);
+
+    /**
+     * Set to true once {@code .concel()} is called in the subscription of the downstream subscriber
+     */
+    private final AtomicBoolean isCancelled = new AtomicBoolean(false);
+
+    /**
+     * Future to track the status of the upstreamResponseTransformer. Will be completed when the future returned by calling
+     * {@code prepare()} on the upstreamResponseTransformer itself completes.
+     */
+    private final CompletableFuture<ResultT> resultFuture;
+
+    /**
+     * The buffer size used to buffer the content received from the downstream subscriber
+     */
+    private final long maximumBufferInBytes;
+
+    /**
+     * This publisher is used to send the bytes received from the downstream subscriber's transformers to a
+     * {@link DelegatingBufferingSubscriber} that will buffer a number of bytes up to {@code maximumBufferSize}.
+     */
+    private final SimplePublisher<ByteBuffer> publisherToUpstream = new SimplePublisher<>();
+
+    /**
+     * The downstream subscriber that is subscribed to this publisher.
+     */
+    private Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber;
+
+    /**
+     * The amount requested by the downstream subscriber that is still left to fulfill. Updated. when the
+     * {@link Subscription#request(long) request} method is called on the downstream subscriber's subscription. Corresponds to the
+     * number of {@code AsyncResponseTransformer} that will be published to the downstream subscriber.
+     */
+    private final AtomicLong outstandingDemand = new AtomicLong(0);
+
+    /**
+     * This flag stops the current thread from publishing transformers while another thread is already publishing.
+     */
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+
+    private SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                 Long maximumBufferSizeInBytes,
+                                 CompletableFuture<ResultT> resultFuture) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(
+            upstreamResponseTransformer, "upstreamResponseTransformer");
+        this.resultFuture = Validate.paramNotNull(
```

Comment:
If the result future is cancelled, we need to propagate the cancellation signal and cancel the request.

Reply:
done, also see line 278